src/core/thread/mlr_manager.cpp

```diff
@@ -81,6 +83,126 @@ void MlrManager::HandleBackboneRouterPrimaryUpdate(BackboneRouter::Leader::State
+            IsAddressMlrRegisteredByAnyChild(addr.GetAddress()))
+        {
+            addr.SetMlrState(kMlrStateRegistered);
+        }
+    }
+#endif // OPENTHREAD_CONFIG_TMF_PROXY_MLR_ENABLE
+
+    ScheduleSend(0);
+}
+
+bool MlrManager::IsAddressMlrRegisteredByNetif(const Ip6::Address &aAddress)
+{
+    bool ret = false;
+
+    for (const Ip6::ExternalNetifMulticastAddress &addr : Get<ThreadNetif>().IterateExternalMulticastAddresses())
+    {
+        if (addr.GetAddress().IsMulticastLargerThanRealmLocal() && addr.GetMlrState() == kMlrStateRegistered &&
+            addr.GetAddress() == aAddress)
+        {
+            ExitNow(ret = true);
+        }
+    }
+
+exit:
+    return ret;
+}
+#endif // OPENTHREAD_CONFIG_MLR_ENABLE
+
+#if OPENTHREAD_CONFIG_TMF_PROXY_MLR_ENABLE
+
+bool MlrManager::IsAddressMlrRegisteredByAnyChildExcept(const Ip6::Address &aAddress, const Child *aExceptChild)
+{
+    bool ret = false;
+
+    for (Child &child : Get<ChildTable>().Iterate(Child::kInStateValidOrRestoring))
+    {
+        if (&child != aExceptChild && child.HasMlrRegisteredAddress(aAddress))
+        {
+            ExitNow(ret = true);
+        }
+    }
+
+exit:
+    return ret;
+}
+
+void MlrManager::UpdateProxiedSubscriptions(Child &             aChild,
+                                            const Ip6::Address *aRegisteredAddresses,
+                                            uint16_t            aRegisteredAddressNum)
+{
+    Ip6::Address address;
+
+    VerifyOrExit(aChild.IsStateValidOrRestoring(), OT_NOOP);
+
+    // Search for the new multicast addresses and set it's flag accordingly
+    for (Child::Ip6AddressIterator iter; aChild.GetNextIp6Address(iter, address) == OT_ERROR_NONE;)
+    {
+        bool isMlrRegistered = false;
+
+        if (!address.IsMulticastLargerThanRealmLocal())
+        {
+            continue;
+        }
+
+        // Check if it's a new multicast address against old addresses
+        for (size_t i = 0; i < aRegisteredAddressNum; i++)
+        {
+            if (aRegisteredAddresses[i] == address)
+            {
+                isMlrRegistered = true;
+                break;
+            }
+        }
+
+#if OPENTHREAD_CONFIG_MLR_ENABLE
+        // Check if it's a new multicast address against parent Netif
+        isMlrRegistered = isMlrRegistered || IsAddressMlrRegisteredByNetif(address);
+#endif
+        // Check if it's a new multicast address against other Children
+        isMlrRegistered = isMlrRegistered || IsAddressMlrRegisteredByAnyChildExcept(address, &aChild);
+
+        // Set MLR state accordingly
+        aChild.SetAddressMlrState(iter.Get(), isMlrRegistered ? kMlrStateRegistered : kMlrStateToRegister);
+    }
+
+exit:
+    if (aChild.HasAnyMlrToRegisterAddress())
+    {
+        LogMulticastAddresses();
+
+        static_assert(Mle::kParentAggregateDelay > 1, "kParentAggregateDelay should be larger than 1 second");
```

Comment:
`kParentAggregateDelay` is a constant, if check, better to put it to the header file `mle_types.hpp` where it is defined?

Reply:
Sounds reasonable. 
But here is the actual logic that imposes such restriction. So, putting it here might be more logical?

Reply:
You put the other check in header file, any rules to follow?
https://github.com/openthread/openthread/blob/d14a14a494c4ca2379ec61a919e2c8c7335ca637/src/core/thread/mle_types.hpp#L266

Reply:
`static_assert(kMlrTimeoutDefault >= kMlrTimeoutMin` is now put under definition of the two constants, which I think is reasonable. 

For `Mle::kParentAggregateDelay`, defining it to 1 may seem fine. If we put the `static_assert(Mle::kParentAggregateDelay > 1)` right after its definition, we don't have the context of why. 

So, I think it's better to be put where `Mle::kParentAggregateDelay > 1` is actually required, which is the statement:
```c++
Random::NonCrypto::GetUint16InRange(1, Mle::kParentAggregateDelay)
```

Thoughts?

Reply:
No strong preference, but may suggest to put it to the header file to avoid repeated check in other components as whenever using it, here in MLR, or in DUA as well, it should be a similar random generation statement.

> `PARENT_AGGREGATE_DELAY` The longest time period that a Parent Router can wait to collect multicast subscriptions and/or DUA registrations from its MTD Children before registering these to the BBR Server

Reply:
moved